pennylane/optimize/adaptive.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Adaptive optimizer"""
+# pylint: disable= no-value-for-parameter
+import pennylane as qml
+
+from pennylane import numpy as np
+
+
+@qml.qfunc_transform
+def append_gate(tape, params, gates):
+    """Append parameterized gates to an existing tape.
+
+    Args:
+        tape (QuantumTape): quantum tape to transform by adding gates
+        params (array[float]): parameters of the gates to be added
+        gates (list[Operator]): list of the gates to be added
+    """
+    for o in tape.operations:
+        qml.apply(o)
+
+    for i, g in enumerate(gates):
+        g.data[0] = params[i]
+        qml.apply(g)
+
+    for m in tape.measurements:
+        qml.apply(m)
+
+
+class AdaptiveOptimizer:
+    r"""Adaptive optimizer.
+
+    Quantum circuits can be built by adding gates
+    `adaptively <https://www.nature.com/articles/s41467-019-10988-2>`_. The adaptive optimizer
+    implements an algorithm that grows and optimizes an input quantum circuit by selecting and
+    adding gates from a user-defined collection of operators. The algorithm starts with adding all
+    the gates to the circuit and computing the circuit gradients with respect to the gate
+    parameters. The algorithm then retains the gate which has the largest gradient and optimizes its
+    parameter. The processes of growing the circuit is repeated until the computed gradients
+    converge within a given threshold. The adaptive optimizer can be used to implement algorithms
+    such as `ADAPT-VQE <https://www.nature.com/articles/s41467-019-10988-2>`_.
+
+    Args:
+        paramopt_steps (float): number of steps for optimizing the parameter of a selected gate
```

Comment:
A very minor comment, but `paramopt_steps` is hard to say out loud 😆 Would `param_steps` also work? The `opt` is maybe redundant.